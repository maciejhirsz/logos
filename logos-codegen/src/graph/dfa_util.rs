use std::{collections::HashSet, iter};

use regex_automata::{
    dfa::{dense::DFA, Automaton},
    util::primitives::StateID,
};

use crate::leaf::LeafId;

pub type OwnedDFA = DFA<Vec<u32>>;

/// Returns an iterator over the matching patterns of a given dfa state. Returns leaf ids in
/// ascending order.
pub fn iter_matches<'a>(state_id: StateID, dfa: &'a OwnedDFA) -> impl Iterator<Item = LeafId> + 'a {
    let num_matches = if dfa.is_match_state(state_id) {
        dfa.match_len(state_id)
    } else {
        0
    };

    (0..num_matches).map(move |match_idx| {
        let pattern_id = dfa.match_pattern(state_id, match_idx);
        LeafId::from(pattern_id)
    })
}

/// Returns an iterator over the child states of a given dfa state. Returns children in order of
/// input byte (0..=255), then eoi. No deduplication of child states is performed.
pub fn iter_children<'a>(dfa: &'a OwnedDFA, state: StateID) -> impl Iterator<Item = StateID> + 'a {
    (0..=u8::MAX)
        .map(move |byte| dfa.next_state(state, byte))
        .chain(iter::once(dfa.next_eoi_state(state)))
}

/// This utility function returns every state accessible by the dfa
/// from a root state. Returns the states in ascending order.
pub fn get_states(dfa: &OwnedDFA, root: StateID) -> impl Iterator<Item = StateID> {
    let mut states = HashSet::new();
    states.insert(root);
    let mut explore_stack = vec![root];
    while let Some(state) = explore_stack.pop() {
        for child in iter_children(dfa, state) {
            if states.insert(child) {
                explore_stack.push(child);
            }
        }
    }

    let mut sorted = states.into_iter().collect::<Vec<_>>();
    sorted.sort_unstable();
    sorted.into_iter()
}
