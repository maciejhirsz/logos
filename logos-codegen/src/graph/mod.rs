use std::ascii::escape_default;
use std::fmt;
use std::{collections::{hash_map::Entry, HashMap, VecDeque}, iter, ops::RangeInclusive};

use regex_automata::{dfa::{dense::DFA, Automaton, StartKind}, nfa::thompson::NFA, util::primitives::StateID, Anchored, MatchKind};

use crate::leaf::{Leaf, LeafId};

mod export;

#[derive(Debug)]
pub struct Config {
    pub prio_over_length: bool,
}

impl Default for Config {
    fn default() -> Self {
        Config { prio_over_length: false }
    }
}

/// Disambiguation error during the attempt to merge two leaf
/// nodes with the same priority
#[derive(Clone, Debug)]
pub struct DisambiguationError(pub Vec<LeafId>);

type OwnedDFA = DFA<Vec<u32>>;

fn iter_matches<'a>(state_id: StateID, dfa: &'a OwnedDFA) -> impl Iterator<Item=LeafId> + 'a {
    let num_matches = if dfa.is_match_state(state_id) {
        dfa.match_len(state_id)
    } else {
        0
    };

    (0..num_matches).map(move |match_idx| {
        let pattern_id = dfa.match_pattern(state_id, match_idx);
        LeafId::from(pattern_id)
    })
}

#[derive(Clone, Copy, Debug, Default)]
pub enum StateType {
    #[default]
    Normal,
    Accept(LeafId),
}

#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq)]
pub struct State {
    pub dfa_id: StateID,
    pub context: Option<LeafId>,
}

impl State {
    fn filter_state_type(&self, state_type: StateType, graph: &Graph) -> StateType {
        if let StateType::Accept(accept_leaf_id) = state_type {
            if let Some(current_leaf_id) = self.context {
                let accept_prio = graph.leaves[accept_leaf_id.0].priority;
                let current_prio = graph.leaves[current_leaf_id.0].priority;
                if accept_prio < current_prio {
                    return StateType::Normal;
                }
            }
        }

        state_type
    }
}

#[derive(Debug, Default)]
pub struct StateData {
    pub state_type: StateType,
    pub normal: Vec<(ByteClass, State)>,
    pub eoi: Option<State>,
}

impl StateData {
    fn iter_children<'a>(&'a self) -> impl Iterator<Item=State> + 'a {
        self.normal.iter().map(|(_bc, s)| s.clone()).chain(self.eoi.iter().cloned())
    }
}


#[derive(Debug)]
pub struct ByteClass {
    pub ranges: Vec<RangeInclusive<u8>>,
}

impl ByteClass {
    fn new() -> Self {
        ByteClass { ranges: Vec::new() }
    }

    fn add_byte(&mut self, byte: u8) {
        if let Some(last) = self.ranges.last_mut() {
            if last.end() + 1 == byte {
                *last = *last.start()..=byte;
                return
            }
        }
        self.ranges.push(byte..=byte);
    }
}

impl fmt::Display for ByteClass {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        for (idx, range) in self.ranges.iter().enumerate() {
            if range.start() == range.end() {
                write!(f, "{}", escape_default(*range.start()))?;
            } else {
                write!(f, "{}..={}", escape_default(*range.start()), escape_default(*range.end()))?;
            }

            if idx < self.ranges.len() - 1 {
                write!(f, "|")?;
            }
        }

        Ok(())
    }
}

#[derive(Debug)]
pub struct Graph {
    config: Config,
    leaves: Vec<Leaf>,
    dfa: OwnedDFA,
    edges: HashMap<State, StateData>,
    root: State,
    errors: Vec<DisambiguationError>,
}

#[derive(Debug)]
struct GraphTraverse {
    state_types: HashMap<StateID, StateType>,
    visit_stack: Vec<State>,
}

impl GraphTraverse {
    fn from_root(root: State) -> Self {
        Self { state_types: HashMap::new(), visit_stack: vec![root] }
    }

    fn get_state_type(&mut self, state_id: StateID, graph: &mut Graph) -> StateType {
        let vacant = match self.state_types.entry(state_id) {
            Entry::Occupied(occupied) => {
                return *occupied.get()
            },
            Entry::Vacant(vacant) => {
                vacant
            },
        };

        let matching_leaves = iter_matches(state_id, &graph.dfa).map(|leaf_id| (leaf_id, graph.leaves[leaf_id.0].priority)).collect::<Vec<_>>();

        let state_type = if let Some(&(highest_leaf_id, highest_priority)) = matching_leaves.iter().max_by_key(|(_leaf_id, priority)| priority) {
            let matching_prio_leaves: Vec<LeafId> = matching_leaves.into_iter().filter(|(leaf_id, priority)| *priority == highest_priority).map(|(leaf_id, _priority)| leaf_id).collect();
            if matching_prio_leaves.len() > 1 {
                graph.errors.push(DisambiguationError(matching_prio_leaves))
            }

            StateType::Accept(highest_leaf_id)
        } else {
            StateType::Normal
        };

        *vacant.insert(state_type)
    }
}

impl Graph {
    pub fn root(&self) -> State {
        self.root
    }

    pub fn get_states<'a>(&'a self) -> impl Iterator<Item=State> + 'a {
        self.edges.keys().cloned()
    }

    pub fn get_state_data(&self, state: &State) -> &StateData {
        self.edges.get(state).expect("Reached unreachable state")
    }

    pub fn leaves(&self) -> &Vec<Leaf> {
        &self.leaves
    }

    pub fn dfa(&self) -> &OwnedDFA {
        &self.dfa
    }

    pub fn errors<'b>(&'b self) -> impl Iterator<Item=DisambiguationError> + 'b {
        self.errors.iter().cloned()
    }

    pub fn new(leaves: Vec<Leaf>, config: Config) -> Result<Self, String> {
        let hirs = leaves.iter().map(|leaf| leaf.pattern.hir()).collect::<Vec<_>>();

        // TODO: utf8 mode here
        let nfa_config = NFA::config()
            .shrink(true);
        let nfa = NFA::compiler().configure(nfa_config).build_many_from_hir(&hirs).map_err(|err| format!("{}", err))?;
        if nfa.has_empty() {
            // TODO Better error handling
            return Err(String::from("Regex includes a zero length match"));
        }
        let dfa_config = DFA::config()
            .accelerate(false)
            .byte_classes(false)
            .minimize(true)
            .match_kind(MatchKind::All)
            .start_kind(StartKind::Anchored);
        let dfa = DFA::builder().configure(dfa_config).build_from_nfa(&nfa).map_err(|err| format!("{}", err))?;


        let start_id = dfa.universal_start_state(Anchored::Yes)
            .expect("Lookaround assertions are disabled, so there should be a universal start state");
        let root = State { dfa_id: start_id, context: None };

        let mut graph = Self { leaves, dfa, edges: HashMap::new(), root, errors: Vec::new(), config };
        let mut traverse = GraphTraverse::from_root(root);

        while let Some(state) = traverse.visit_stack.pop() {
            if graph.edges.contains_key(&state) { continue }
            let state_data = graph.gen_state_data(state, &mut traverse);
            traverse.visit_stack.extend(state_data.iter_children());
            graph.edges.insert(state, state_data);
        }

        // TODO: prune nodes that don't lead to any more accept states before reaching the dead
        // node (0)

        Ok(graph)
    }

    fn gen_state_data(&mut self, state: State, traverse: &mut GraphTraverse) -> StateData {
        let state_type = state.filter_state_type(traverse.get_state_type(state.dfa_id, self), self);

        let mut result: HashMap<State, ByteClass> = HashMap::new();
        for input_byte in u8::MIN..=u8::MAX {
            let next_id = self.dfa.next_state(state.dfa_id, input_byte);

            // Don't need to account for the dead state
            if next_id.as_usize() == 0 { continue }

            let next_state = self.propagate_context(state, next_id, traverse);

            let bytes_to_next_state = result.entry(next_state).or_insert(ByteClass::new());
            bytes_to_next_state.add_byte(input_byte);
        }

        let normal = result.into_iter().map(|(s, bc)| (bc, s)).collect();

        let eoi_id  = self.dfa.next_eoi_state(state.dfa_id);
        let eoi = if eoi_id.as_usize() == 0 {
            None
        } else {
            Some(self.propagate_context(state, eoi_id, traverse))
        };

        StateData {
            state_type,
            normal,
            eoi,
        }
    }

    fn propagate_context(&mut self, prev: State, next_id: StateID, traverse: &mut GraphTraverse) -> State {
        let mut next_state_type = traverse.get_state_type(next_id, self);

        if self.config.prio_over_length {
            next_state_type = prev.filter_state_type(next_state_type, self);
        };

        let context = match next_state_type {
            StateType::Normal => prev.context,
            StateType::Accept(leaf_id) => Some(leaf_id),
        };

        State { dfa_id: next_id, context }
    }

}
