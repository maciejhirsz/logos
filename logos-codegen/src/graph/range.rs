use regex_syntax::hir::ClassBytesRange;
use regex_syntax::hir::ClassUnicodeRange;
use regex_syntax::utf8::Utf8Range;

use std::cmp::{Ord, Ordering};

#[derive(Clone, Copy, PartialEq, Eq, Hash)]
pub struct Range {
    pub start: u8,
    pub end: u8,
}

impl Range {
    pub fn as_byte(&self) -> Option<u8> {
        if self.is_byte() {
            Some(self.start)
        } else {
            None
        }
    }

    pub fn is_byte(&self) -> bool {
        self.start == self.end
    }
}

impl From<u8> for Range {
    fn from(byte: u8) -> Range {
        Range {
            start: byte,
            end: byte,
        }
    }
}

impl From<&u8> for Range {
    fn from(byte: &u8) -> Range {
        Range::from(*byte)
    }
}

impl Iterator for Range {
    type Item = u8;

    fn next(&mut self) -> Option<u8> {
        match self.start.cmp(&self.end) {
            std::cmp::Ordering::Less => {
                let res = self.start;
                self.start += 1;

                Some(res)
            }
            std::cmp::Ordering::Equal => {
                let res = self.start;

                // Necessary so that range 0xFF-0xFF doesn't loop forever
                self.start = 0xFF;
                self.end = 0x00;

                Some(res)
            }
            std::cmp::Ordering::Greater => None,
        }
    }
}

impl PartialOrd for Range {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Range {
    fn cmp(&self, other: &Self) -> Ordering {
        self.start.cmp(&other.start)
    }
}

impl From<Utf8Range> for Range {
    fn from(r: Utf8Range) -> Range {
        Range {
            start: r.start,
            end: r.end,
        }
    }
}

impl From<ClassUnicodeRange> for Range {
    fn from(r: ClassUnicodeRange) -> Range {
        let start = r.start() as u32;
        let end = r.end() as u32;

        if start >= 128 || end >= 128 && end != 0x0010FFFF {
            panic!("Casting non-ascii ClassUnicodeRange to Range")
        }

        Range {
            start: start as u8,
            end: end as u8,
        }
    }
}

impl From<ClassBytesRange> for Range {
    fn from(r: ClassBytesRange) -> Range {
        Range {
            start: r.start(),
            end: r.end(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn range_iter_one() {
        let byte = Range::from(b'!');
        let collected = byte.take(1000).collect::<Vec<_>>();

        assert_eq!(b"!", &collected[..]);
    }

    #[test]
    fn range_iter_few() {
        let byte = Range {
            start: b'a',
            end: b'd',
        };
        let collected = byte.take(1000).collect::<Vec<_>>();

        assert_eq!(b"abcd", &collected[..]);
    }

    #[test]
    fn range_iter_bunds() {
        let byte = Range::from(0xFA..=0xFF);

        let collected = byte.take(1000).collect::<Vec<_>>();

        assert_eq!(b"\xFA\xFB\xFC\xFD\xFE\xFF", &collected[..]);
    }
}
