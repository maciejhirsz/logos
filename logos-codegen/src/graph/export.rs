use crate::graph::{ByteClass, Graph, State};
use std::fmt::Write;

enum NodeColor {
    Black,
    Green,
}

impl NodeColor {
    fn fmt_dot(&self) -> &'static str {
        match self {
            Self::Black => "black",
            Self::Green => "green",
        }
    }

    fn fmt_mmd(&self) -> &'static str {
        match self {
            Self::Black => "#000000",
            Self::Green => "#00C853",
        }
    }
}

trait ExportFormat {
    fn write_header(s: &mut String) -> std::fmt::Result;

    fn write_footer(s: &mut String) -> std::fmt::Result;

    fn write_node(s: &mut String, id: &str, label: &str, color: NodeColor) -> std::fmt::Result;

    fn write_link(s: &mut String, from: &str, to: &str, label: &str) -> std::fmt::Result;

    fn fmt_range(bc: &ByteClass) -> String;
}

struct Dot;

impl ExportFormat for Dot {
    fn write_header(s: &mut String) -> std::fmt::Result {
        writeln!(s, "digraph {{")?;
        writeln!(s, "node[shape=box];")?;
        writeln!(s, "splines=ortho;")
    }

    fn write_footer(s: &mut String) -> std::fmt::Result {
        writeln!(s, "}}")
    }

    fn write_node(s: &mut String, id: &str, label: &str, color: NodeColor) -> std::fmt::Result {
        writeln!(s, "{id}[label=\"{label}\",color={}];", color.fmt_dot())
    }

    fn write_link(s: &mut String, from: &str, to: &str, label: &str) -> std::fmt::Result {
        writeln!(s, "{from}->{to} [label=\"{label}\"];")
    }

    fn fmt_range(bc: &ByteClass) -> String {
        bc.to_string().escape_default().to_string()
    }
}

struct Mermaid;

impl ExportFormat for Mermaid {
    // TODO: This should really use the mermaid diagram type
    // `stateDiagram` instead. since it more closely aligns with what we are representing.
    fn write_header(s: &mut String) -> std::fmt::Result {
        writeln!(s, "flowchart TB")
    }

    fn write_footer(_s: &mut String) -> std::fmt::Result {
        Ok(())
    }

    fn write_node(s: &mut String, id: &str, label: &str, color: NodeColor) -> std::fmt::Result {
        writeln!(s, "{id}[\"{label}\"]")?;
        writeln!(s, "style {id} stroke:{}", color.fmt_mmd())
    }

    fn write_link(s: &mut String, from: &str, to: &str, label: &str) -> std::fmt::Result {
        writeln!(s, "{from}-->|\"{label}\"|{to}")
    }

    fn fmt_range(bc: &ByteClass) -> String {
        let mut result = String::new();
        for c in bc.to_string().chars() {
            match c {
                '"' => {
                    let _ = result.write_str("&quot");
                }
                '\\' => {
                    let _ = result.write_str("\\\\");
                }
                _ => result.push(c),
            }
        }

        result
    }
}

impl Graph {
    /// Writes the `Graph` to a dot file.
    pub fn get_dot(&self) -> Result<String, std::fmt::Error> {
        self.export_graph::<Dot>()
    }

    /// Writes the `Graph` to a mermaid file.
    pub fn get_mermaid(&self) -> Result<String, std::fmt::Error> {
        self.export_graph::<Mermaid>()
    }

    fn export_graph<Fmt: ExportFormat>(&self) -> Result<String, std::fmt::Error> {
        fn format_state(state: &State, fancy: bool) -> String {
            let state_id = state.0;
            if fancy {
                format!("State {state_id}")
            } else {
                format!("n{state_id}")
            }
        }

        let mut s = String::new();

        Fmt::write_header(&mut s)?;

        for state in self.iter_states() {
            let data = self.get_state(state);

            let id = format_state(&state, false);
            let label = format_state(&state, true);
            let color = if data.state_type.early.is_some() || data.state_type.accept.is_some() {
                NodeColor::Green
            } else {
                NodeColor::Black
            };

            Fmt::write_node(&mut s, &id, &label, color)?;

            for (bc, to_state) in &data.normal {
                let to_id = format_state(to_state, false);
                let range = Fmt::fmt_range(bc);
                Fmt::write_link(&mut s, &id, &to_id, &range)?;
            }
            if let Some(eoi_state) = &data.eoi {
                let eoi_id = format_state(eoi_state, false);
                Fmt::write_link(&mut s, &id, &eoi_id, "EOI")?;
            }
        }

        Fmt::write_footer(&mut s)?;

        Ok(s)
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;
    use proc_macro2::Span;

    use crate::{graph::Config, leaf::Leaf, pattern::Pattern};

    use super::*;

    #[test]
    fn range_fmt_single_ascii_byte() {
        let r = ByteClass {
            ranges: vec![0x6C..=0x6C],
        };
        assert_snapshot!(Dot::fmt_range(&r), @"l");
        assert_snapshot!(Mermaid::fmt_range(&r), @"l");
    }

    #[test]
    fn range_fmt_ascii_bytes() {
        let r = ByteClass {
            ranges: vec![0x61..=0x7A],
        };
        assert_snapshot!(Dot::fmt_range(&r), @"a..=z");
        assert_snapshot!(Mermaid::fmt_range(&r), @"a..=z");
    }

    #[test]
    fn range_fmt_single_escaped_ascii_byte() {
        let r = ByteClass {
            ranges: vec![0x22..=0x22],
        };
        assert_snapshot!(Dot::fmt_range(&r), @r###"\\\""###);
        assert_snapshot!(Mermaid::fmt_range(&r), @r###"\\&quot"###);

        let r = ByteClass {
            ranges: vec![0x5C..=0x5C],
        };
        assert_snapshot!(Dot::fmt_range(&r), @r###"\\\\"###);
        assert_snapshot!(Mermaid::fmt_range(&r), @r###"\\\\"###);
    }

    #[test]
    fn range_fmt_single_hex_byte() {
        let r = ByteClass {
            ranges: vec![0x0A..=0x0A],
        };
        assert_snapshot!(Dot::fmt_range(&r), @r###"\\n"###);
        assert_snapshot!(Mermaid::fmt_range(&r), @r###"\\n"###);
    }

    #[test]
    fn range_fmt_hex_bytes() {
        let r = ByteClass {
            ranges: vec![0x0A..=0x10],
        };
        assert_snapshot!(Dot::fmt_range(&r), @r###"\\n..=\\x10"###);
        assert_snapshot!(Mermaid::fmt_range(&r), @r###"\\n..=\\x10"###);
    }

    fn export_graphs(patterns: Vec<&str>) -> [String; 2] {
        let leaves = patterns
            .into_iter()
            .map(|src| {
                Leaf::new(
                    Span::call_site(),
                    Pattern::compile(false, src, src.to_string(), true, false)
                        .expect("Unable to compile pattern"),
                )
            })
            .collect();

        let config = Config { utf8_mode: true };
        let graph = Graph::new(leaves, config).expect("Unable to compile graph");
        let dot = graph.export_graph::<Dot>().unwrap();
        let mmd = graph.export_graph::<Mermaid>().unwrap();

        [dot, mmd]
    }

    #[test]
    fn fork() {
        let patterns = vec!["[a-y]", "z"];

        let [dot, mmd] = export_graphs(patterns);
        assert_snapshot!(dot);
        assert_snapshot!(mmd);
    }

    #[test]
    fn rope() {
        let patterns = vec!["rope"];

        let [dot, mmd] = export_graphs(patterns);
        assert_snapshot!(dot);
        assert_snapshot!(mmd);
    }

    #[test]
    fn rope_with_miss_first() {
        let patterns = vec!["f(ee)?"];

        let [dot, mmd] = export_graphs(patterns);
        assert_snapshot!(dot);
        assert_snapshot!(mmd);
    }

    #[test]
    fn rope_with_miss_any() {
        let patterns = vec!["fe{0,2}"];

        let [dot, mmd] = export_graphs(patterns);
        assert_snapshot!(dot);
        assert_snapshot!(mmd);
    }
}
