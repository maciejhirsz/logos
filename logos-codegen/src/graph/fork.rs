use crate::graph::{Disambiguate, Graph, NodeId, Range};

#[derive(Clone)]
pub struct Fork {
    /// LUT matching byte -> node id
    lut: Box<[Option<NodeId>; 256]>,
    /// State to go to if no arms are matching
    pub miss: Option<NodeId>,
}

impl Fork {
    pub fn new() -> Self {
        Fork {
            lut: Box::new([None; 256]),
            miss: None,
        }
    }

    pub fn miss<M>(mut self, miss: M) -> Self
    where
        M: Into<Option<NodeId>>,
    {
        self.miss = miss.into();
        self
    }

    pub fn add_branch<R, T>(&mut self, range: R, then: NodeId, graph: &mut Graph<T>)
    where
        R: Into<Range>,
        T: Disambiguate,
    {
        for byte in range.into() {
            match &mut self.lut[byte as usize] {
                Some(other) if *other != then => {
                    *other = graph.merge(*other, then);
                }
                opt => *opt = Some(then),
            }
        }
    }

    // TODO: Add result with a printable error
    pub fn merge<T>(&mut self, other: Fork, graph: &mut Graph<T>)
    where
        T: Disambiguate,
    {
        self.miss = match (self.miss, other.miss) {
            (None, None) => None,
            (Some(id), None) | (None, Some(id)) => Some(id),
            (Some(a), Some(b)) => Some(graph.merge(a, b)),
        };

        for (left, right) in self.lut.iter_mut().zip(other.lut.iter()) {
            *left = match (*left, *right) {
                (None, None) => continue,
                (Some(id), None) | (None, Some(id)) => Some(id),
                (Some(a), Some(b)) => Some(graph.merge(a, b)),
            }
        }
    }

    pub fn branches(&self) -> ForkIter<'_> {
        ForkIter {
            offset: 0,
            lut: &*self.lut,
        }
    }

    /// Checks if all bytes in the `range` have a branch on this
    /// fork, and those branches are resolve to the same `NodeId`.
    pub fn contains<R>(&self, range: R) -> Option<NodeId>
    where
        R: Into<Range>,
    {
        let mut range = range.into();
        let byte = range.next()?;
        let first = self.lut[byte as usize]?;

        for byte in range {
            if first != self.lut[byte as usize]? {
                return None;
            }
        }

        Some(first)
    }

    pub fn branch<R>(mut self, range: R, then: NodeId) -> Self
    where
        R: Into<Range>,
    {
        for byte in range.into() {
            match &mut self.lut[byte as usize] {
                Some(other) if *other != then => {
                    panic!("Overlapping branches");
                }
                opt => *opt = Some(then),
            }
        }
        self
    }

    pub fn shake<T>(&self, graph: &Graph<T>, filter: &mut [bool]) {
        if let Some(id) = self.miss {
            if !filter[id.get()] {
                filter[id.get()] = true;
                graph[id].shake(graph, filter);
            }
        }

        for (_, id) in self.branches() {
            if !filter[id.get()] {
                filter[id.get()] = true;
                graph[id].shake(graph, filter);
            }
        }
    }
}

pub struct ForkIter<'a> {
    offset: usize,
    lut: &'a [Option<NodeId>; 256],
}

impl<'a> Iterator for ForkIter<'a> {
    type Item = (Range, NodeId);

    fn next(&mut self) -> Option<Self::Item> {
        // Consume empty slots
        self.offset += self.lut[self.offset..]
            .iter()
            .take_while(|next| next.is_none())
            .count();

        let then = self.lut.get(self.offset).copied().flatten()?;
        let start = self.offset;

        // Consume all slots with same NodeId target
        self.offset += self.lut[self.offset..]
            .iter()
            .take_while(|next| **next == Some(then))
            .count();

        Some((
            Range {
                start: start as u8,
                end: (self.offset - 1) as u8,
            },
            then,
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::graph::Node;
    use pretty_assertions::assert_eq;

    #[test]
    fn fork_iter() {
        let mut buf = [None; 256];

        for byte in b'4'..=b'7' {
            buf[byte as usize] = Some(NodeId::new(1));
        }
        for byte in b'a'..=b'd' {
            buf[byte as usize] = Some(NodeId::new(2));
        }

        let iter = ForkIter {
            offset: 0,
            lut: &buf,
        };

        assert_eq!(
            &[
                (
                    Range {
                        start: b'4',
                        end: b'7'
                    },
                    NodeId::new(1)
                ),
                (
                    Range {
                        start: b'a',
                        end: b'd'
                    },
                    NodeId::new(2)
                ),
            ],
            &*iter.collect::<Vec<_>>(),
        );
    }

    #[test]
    fn merge_no_conflict() {
        let mut graph = Graph::new();

        let leaf1 = graph.push(Node::Leaf("FOO"));
        let leaf2 = graph.push(Node::Leaf("BAR"));

        let mut fork = Fork::new().branch(b'1', leaf1);

        fork.merge(Fork::new().branch(b'2', leaf2), &mut graph);

        assert_eq!(fork, Fork::new().branch(b'1', leaf1).branch(b'2', leaf2));
    }

    #[test]
    fn merge_miss_right() {
        let mut graph = Graph::new();

        let leaf1 = graph.push(Node::Leaf("FOO"));
        let leaf2 = graph.push(Node::Leaf("BAR"));

        let mut fork = Fork::new().branch(b'1', leaf1);

        fork.merge(Fork::new().miss(leaf2), &mut graph);

        assert_eq!(fork, Fork::new().branch(b'1', leaf1).miss(leaf2));
    }

    #[test]
    fn merge_miss_left() {
        let mut graph = Graph::new();

        let leaf1 = graph.push(Node::Leaf("FOO"));
        let leaf2 = graph.push(Node::Leaf("BAR"));

        let mut fork = Fork::new().miss(leaf1);

        fork.merge(Fork::new().branch(b'2', leaf2), &mut graph);

        assert_eq!(fork, Fork::new().branch(b'2', leaf2).miss(leaf1));
    }

    #[test]
    fn contains_byte() {
        let fork = Fork::new().branch('a'..='z', NodeId::new(42));

        assert_eq!(fork.contains(b't'), Some(NodeId::new(42)));
    }

    #[test]
    fn contains_range() {
        let fork = Fork::new()
            .branch('a'..='m', NodeId::new(42))
            .branch('n'..='z', NodeId::new(42));

        assert_eq!(fork.contains('i'..='r'), Some(NodeId::new(42)));
        assert_eq!(fork.contains('a'..='z'), Some(NodeId::new(42)));
    }

    #[test]
    fn contains_different_ranges() {
        let fork = Fork::new()
            .branch('a'..='m', NodeId::new(42))
            .branch('n'..='z', NodeId::new(47));

        assert_eq!(fork.contains('i'..='r'), None);
        assert_eq!(fork.contains('a'..='z'), None);
        assert_eq!(fork.contains('d'..='f'), Some(NodeId::new(42)));
        assert_eq!(fork.contains('n'..='p'), Some(NodeId::new(47)));
    }
}
