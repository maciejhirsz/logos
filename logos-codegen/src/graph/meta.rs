use std::cmp::min;
use std::collections::BTreeMap;
use std::ops::{Index, IndexMut};

use crate::graph::{Graph, Node, NodeId};

#[derive(Debug)]
pub struct Meta {
    map: BTreeMap<NodeId, MetaItem>,
}

#[derive(Debug, Default)]
pub struct MetaItem {
    /// Number of references to this node
    pub refcount: usize,
    /// Minimum number of bytes that ought to be read for this
    /// node to find a match
    pub min_read: usize,
    /// Marks whether or not this node leads to a loop entry node.
    pub is_loop_init: bool,
    /// Ids of other nodes that point to this node while this
    /// node is on a stack (creating a loop)
    pub loop_entry_from: Vec<NodeId>,
}

impl Index<NodeId> for Meta {
    type Output = MetaItem;

    fn index(&self, id: NodeId) -> &MetaItem {
        &self.map[&id]
    }
}

impl IndexMut<NodeId> for Meta {
    fn index_mut(&mut self, id: NodeId) -> &mut MetaItem {
        self.map.entry(id).or_default()
    }
}

impl MetaItem {
    fn loop_entry(&mut self, id: NodeId) {
        if let Err(idx) = self.loop_entry_from.binary_search(&id) {
            self.loop_entry_from.insert(idx, id);
        }
    }
}

impl Meta {
    pub fn analyze<T>(root: NodeId, graph: &Graph<T>) -> Self {
        let mut meta = Meta {
            map: Default::default(),
        };

        meta.first_pass(root, root, graph, &mut Vec::new());

        meta
    }

    pub fn first_pass<T>(
        &mut self,
        this: NodeId,
        parent: NodeId,
        graph: &Graph<T>,
        stack: &mut Vec<NodeId>,
    ) -> &MetaItem {
        let meta = &mut self[this];
        let is_done = meta.refcount > 0;

        meta.refcount += 1;

        if stack.contains(&this) {
            meta.loop_entry(parent);
            self[parent].is_loop_init = true;
        }
        if is_done {
            return &self[this];
        }

        stack.push(this);

        let mut min_read;

        match &graph[this] {
            Node::Fork(fork) => {
                min_read = usize::max_value();
                for (_, id) in fork.branches() {
                    let meta = self.first_pass(id, this, graph, stack);

                    if meta.is_loop_init {
                        min_read = 1;
                    } else {
                        min_read = min(min_read, meta.min_read + 1);
                    }
                }
                if let Some(id) = fork.miss {
                    let meta = self.first_pass(id, this, graph, stack);

                    if meta.is_loop_init {
                        min_read = 0;
                    } else {
                        min_read = min(min_read, meta.min_read);
                    }
                }
                if min_read == usize::max_value() {
                    min_read = 0;
                }
            }
            Node::Rope(rope) => {
                min_read = rope.pattern.len();
                let meta = self.first_pass(rope.then, this, graph, stack);

                if !meta.is_loop_init {
                    min_read += meta.min_read;
                }

                if let Some(id) = rope.miss.first() {
                    let meta = self.first_pass(id, this, graph, stack);

                    if meta.is_loop_init {
                        min_read = 0;
                    } else {
                        min_read = min(min_read, meta.min_read);
                    }
                }
            }
            Node::Leaf(_) => min_read = 0,
        }

        stack.pop();

        let meta = &mut self[this];
        meta.min_read = min_read;
        let second_pass = meta.loop_entry_from.clone();

        for id in second_pass {
            self.meta_second_pass(id, graph);
        }

        &self[this]
    }

    fn meta_second_pass<T>(&mut self, id: NodeId, graph: &Graph<T>) {
        let mut min_read;

        match &graph[id] {
            Node::Fork(fork) => {
                min_read = usize::max_value();
                for (_, id) in fork.branches() {
                    let meta = &self[id];

                    if meta.is_loop_init {
                        min_read = 1;
                    } else {
                        min_read = min(min_read, meta.min_read + 1);
                    }
                }
                if min_read == usize::max_value() {
                    min_read = 0;
                }
            }
            Node::Rope(rope) => {
                min_read = rope.pattern.len();
                let meta = &self[rope.then];

                if !meta.is_loop_init {
                    min_read += meta.min_read;
                }
            }
            Node::Leaf(_) => unreachable!(),
        }

        self[id].min_read = min_read;
    }
}
