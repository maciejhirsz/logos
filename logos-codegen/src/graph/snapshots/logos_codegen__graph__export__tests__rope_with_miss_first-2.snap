---
source: logos-codegen/src/graph/export.rs
expression: mmd
---
flowchart TB
n1024ctx0["State 1024 (ctx 0)"]
style n1024ctx0 stroke:#00C853
n1536ctx0["State 1536 (ctx 0)"]
style n1536ctx0 stroke:#00C853
n1536ctx0-->|"e"|n3072ctx0
n2048["State 2048"]
style n2048 stroke:#000000
n2048-->|"f"|n2560
n2560["State 2560"]
style n2560 stroke:#000000
n2560-->|"\\x00..=d|f..=\\xff"|n1024ctx0
n2560-->|"e"|n1536ctx0
n2560-->|"EOI"|n1024ctx0
n3072ctx0["State 3072 (ctx 0)"]
style n3072ctx0 stroke:#000000
n3072ctx0-->|"\\x00..=\\xff"|n1024ctx0
n3072ctx0-->|"EOI"|n1024ctx0
